include/boost/url/impl/decoded_view.hpp

```diff
@@ -0,0 +1,229 @@
+        size_type n,
+        bool plus_to_space,
+        int) noexcept
+        : begin_(str)
+        , pos_(str + n)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+public:
+    using value_type = char;
+    using reference = char;
+    using pointer = void const*;
+    using const_reference = char;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category =
+        std::bidirectional_iterator_tag;
+
+    iterator() = default;
+
+    iterator(iterator const&) = default;
+
+    iterator&
+    operator=(iterator const&) = default;
+
+    reference
+    operator*() const noexcept
+    {
+        char c{'\0'};
+        pct_decode_opts opt;
+        opt.plus_to_space = plus_to_space_;
+        pct_decode_unchecked(
+            &c, &c+1, string_view(pos_, end_), opt);
+        return c;
+    }
+
+    iterator&
+    operator++() noexcept
+    {
+        BOOST_ASSERT(pos_ != nullptr);
+        if (*pos_ != '%')
+            ++pos_;
+        else
+            pos_ += 3;
+        return *this;
+    }
+
+    iterator&
+    operator--() noexcept
+    {
+        BOOST_ASSERT(pos_ != begin_);
+        if (pos_ - begin_ < 3 ||
+            pos_[-3] != '%')
+            --pos_;
+        else
+            pos_ -= 3;
+        return *this;
+    }
+
+    iterator
+    operator++(int) noexcept
+    {
+        auto tmp = *this;
+        ++*this;
+        return tmp;
+    }
+
+    iterator
+    operator--(int) noexcept
+    {
+        auto tmp = *this;
+        --*this;
+        return tmp;
+    }
+
+    const char*
+    base()
+    {
+        return pos_;
+    }
+
+    bool
+    operator==(
+        iterator const& other) const noexcept
+    {
+        return pos_ == other.pos_;
+    }
+
+    bool
+    operator!=(
+        iterator const& other) const noexcept
+    {
+        return !(*this == other);
+    }
+};
+
+inline
+auto
+decoded_view::
+begin() const noexcept ->
+    decoded_view::const_iterator
+{
+    return {p_, n_, plus_to_space_};
+}
+
+inline
+auto
+decoded_view::
+end() const noexcept ->
+    decoded_view::const_iterator
+{
+    return {p_, n_, plus_to_space_, 0};
+}
+
+inline
+decoded_view::const_reference
+decoded_view::
+front() const
+{
+    BOOST_ASSERT( !empty() );
+    return *begin();
+}
+
+inline
+decoded_view::const_reference
+decoded_view::
+back() const
+{
+    BOOST_ASSERT( !empty() );
+    return *--end();
+}
+
+namespace detail
+{
+template<class C, class I, class = void>
+struct has_assign_from : std::false_type {};
+
+template<class C, class I>
+struct has_assign_from<C, I, boost::void_t<
+    decltype(
+        std::declval<C&>()
+            .assign(
+                std::declval<I>(),
+                std::declval<I>())
+                ) > > : std::true_type
+{
+};
+
+template<class C, class I, class = void>
+struct has_append_from : std::false_type {};
+
+template<class C, class I>
+struct has_append_from<C, I, boost::void_t<
```

Comment:
I think we might roll these both up into a single trait that checks for both, and also checks that `value_type` is `char`. And we might call it `is_mutable_string`. I would use this in Boost.HTTP.Proto, so we might consider making it public and putting it in `urls::grammar`

Reply:
Hmm I am not sure, maybe start there and if we build up more algorithms then we might rename it. For example we could do free-function versions of `append_to` and `assign_to` which accept an InputIterator range of char and outputs to a MutableString

Reply:
> maybe start there

Do you mean "start there" as in "leave it as it is until we have a use case for a public trait" or "maybe start there" as in "create the public trait so maybe we can come up with algorithms that use it"?

Reply:
Start with it in grammar/is_mutable_string.hpp